haskell/private/actions/link.bzl

```diff
@@ -344,18 +524,53 @@ def _add_external_libraries(args, ext_libs):
       args: Args object.
       ext_libs: external_libraries from HaskellBuildInfo
     """
-    seen_libs = set.empty()
-    for ext_lib in set.to_list(ext_libs):
-        lib = ext_lib.mangled_lib
-        lib_name = get_lib_name(lib)
-        if not set.is_member(seen_libs, lib_name):
-            set.mutable_insert(seen_libs, lib_name)
-            args.add_all([
-                "-l{0}".format(lib_name),
-                "-L{0}".format(paths.dirname(lib.path)),
-            ])
 
-def _separate_static_and_dynamic_libraries(ext_libs, dynamic):
+    # Normally, GHC adds `-Wl,-rpath,<dir>` to the linker for every
+    # `<dir>` that is added with `-L` (`-dynload=sysdep`).
+    # These paths are of no value to bazel, because the `lib.path`
+    # dirs we add below are only valid at link-time. We add the
+    # correct load-time paths (`lib.short_path`) below manually,
+    # and use `-dynload=deploy` so GHC doesn’t add any.
+    # See https://downloads.haskell.org/~ghc/7.6.1/docs/html/users_guide/using-shared-libs.html#finding-shared-libs-unix
+    # The difference between link-time and load-time comes from
+    # bazel sandboxing the runtime.
+    # (see: bazel runfiles).
+    args.add("-dynload=deploy")
```

Comment:
Unfortunately, this breaks dependencies on prebuilt packages or the rts on MacOS. E.g. `//tests/binary-dynamic` fails, because it depends on `@rpath/libHSbase-4.12.0.0-ghc8.6.3.dylib`, `@rpath/libHSghc-prim-0.5.3-ghc8.6.3.dylib`, `@rpath/libHSrts_thr-ghc8.6.3.dylib`, but there's no `rpath` entry for any of them.

I think this requires additional work to forward `library-dirs` fields of prebuilt dependencies and add them to `rpath`. But, maybe that's better done in a separate PR, and `-dynload=deploy` is dropped for now.

Reply:
removed it from this PR.